chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,637 @@
+}
+
+/// The reward is distributed among delegations proportionally to their balance
+fn distribute_delegations_pos_reward<P: PoSAccountingView>(
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    delegation_shares: &BTreeMap<DelegationId, Amount>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_delegations_balance: Amount,
+    total_delegations_reward: Amount,
+) -> Result<(Vec<PoSAccountingUndo>, Amount), ConnectTransactionError> {
+    let rewards_per_delegation = calculate_rewards_per_delegation(
+        delegation_shares,
+        pool_id,
+        total_delegations_balance,
+        total_delegations_reward,
+    )?;
+
+    // increase the delegation balances
+    let delegation_undos = rewards_per_delegation
+        .iter()
+        .map(|(delegation_id, reward)| {
+            accounting_adapter
+                .operations(TransactionSource::Chain(block_id))
+                .delegate_staking(*delegation_id, *reward)
+                .map_err(ConnectTransactionError::PoSAccountingError)
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // Due to integer arithmetics there can be a small remainder after all the delegations distributed.
+    // This remainder goes to the pool's owner
+    let total_delegations_reward_distributed =
+        rewards_per_delegation.iter().try_fold(Amount::ZERO, |acc, (_, v)| {
+            (acc + *v).ok_or(ConnectTransactionError::DelegationsRewardSumFailed(
+                block_id, pool_id,
+            ))
+        })?;
+
+    let delegations_reward_remainder =
+        (total_delegations_reward - total_delegations_reward_distributed).ok_or(
+            ConnectTransactionError::DistributedDelegationsRewardExceedTotal(
+                pool_id,
+                block_id,
+                total_delegations_reward_distributed,
+                total_delegations_reward,
+            ),
+        )?;
+    Ok((delegation_undos, delegations_reward_remainder))
+}
+
+fn calculate_rewards_per_delegation(
+    delegation_shares: &BTreeMap<DelegationId, Amount>,
+    pool_id: PoolId,
+    total_delegations_amount: Amount,
+    total_delegations_reward_amount: Amount,
+) -> Result<Vec<(DelegationId, Amount)>, ConnectTransactionError> {
+    assert_eq_size!(Amount, Uint128);
+    ensure!(
+        total_delegations_amount != Amount::ZERO,
+        ConnectTransactionError::TotalDelegationBalanceZero(pool_id)
+    );
+
+    let total_delegations_balance = Uint256::from_amount(total_delegations_amount);
+    let total_delegations_reward = Uint256::from_amount(total_delegations_reward_amount);
+    delegation_shares
+        .iter()
+        .map(
+            |(delegation_id, balance_amount)| -> Result<_, ConnectTransactionError> {
+                let balance = Uint256::from_amount(*balance_amount);
+                let reward = (total_delegations_reward * balance) / total_delegations_balance;
+                let reward: UnsignedIntType = reward.try_into().map_err(|_| {
```

Comment:
It's better to make this `amount::UnsignedIntType` to make it readable that this is about amount.